### PR TITLE
Swarm tests and bug fixes

### PR DESCRIPTION
**Pull request by marklysze**
## Why are these changes needed?

Implementation of tests for swarms.

Includes a number of bug fixes for swarms as well:
- Multiple tool calling where context variables are updated in each
- Matching name for the initial message

Added a number of type checks as well.

## Checks

- [ ] I've included any doc changes needed for https://ag2ai.github.io/ag2/. See https://ag2ai.github.io/ag2/docs/Contribute#documentation to build and test documentation locally.
- [ ] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [X] I've made sure all auto checks have passed.

### CODE REVIEW

File: autogen/agentchat/contrib/swarm_agent.py [hunk 1]
```diff
@@ -340,8 +363,18 @@ def generate_swarm_tool_reply(
...
+
+            # Loop through tool calls individually (so context can be updated after each function call)
+            next_agent = None
+            tool_responses_inner = []
+            contents = []
+            for index in range(tool_calls):
+
+                # 1. add context_variables to the tool call arguments
+                tool_call = message["tool_calls"][index]
+
```
Comment (yiranwu0) [inline autogen/agentchat/contrib/swarm_agent.py:377]: Not sure if you tested the original code, I think the context variables are already being updated after each function call. The context variable dict is a reference.

File: test/agentchat/contrib/test_swarm.py [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+    with pytest.raises(ValueError):
+        AFTER_WORK("INVALID_OPTION")
+
+
+def test_on_condition():
+    """Test ON_CONDITION initialization"""
+    agent = SwarmAgent("test")
+    condition = ON_CONDITION(agent=agent, condition="test condition")
+    assert condition.agent == agent
+    assert condition.condition == "test condition"
```
Comment (yiranwu0) [inline test/agentchat/contrib/test_swarm.py:85]: I feel like this is excessive testing

File: test/agentchat/contrib/test_swarm.py [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+    """Test SwarmAgent initialization with valid and invalid parameters"""
+
+    # Valid initialization
+    agent = SwarmAgent("test_agent")
+    assert agent.name == "test_agent"
+    assert agent.human_input_mode == "NEVER"
+
+    # Invalid functions parameter
+    with pytest.raises(TypeError):
+        SwarmAgent("test_agent", functions="invalid")
```
Comment (yiranwu0) [inline test/agentchat/contrib/test_swarm.py:32]: Don't think this is needed.

File: test/agentchat/contrib/test_swarm.py [hunk 4]
```diff
@@ -0,0 +1,444 @@
...
+    )
+
+    assert chat_result.chat_history[0].get("name") == "MyUser"  # Should persist
+    assert chat_result.chat_history[1].get("name") == "SecondAgent"
+
+
+def test_swarm_transitions():
+    """Test different swarm transition scenarios"""
+    agent1 = SwarmAgent("agent1")
+    agent2 = SwarmAgent("agent2")
```
Comment (yiranwu0) [inline test/agentchat/contrib/test_swarm.py:144]: Does this use OpenAI?

File: test/agentchat/contrib/test_swarm.py [hunk 5]
```diff
@@ -0,0 +1,444 @@
...
+    multiple_messages = [
+        {"role": "user", "content": "First message"},
+        {"role": "assistant", "name": "agent2", "content": "Response"},
+    ]
+
+    chat_result, context_vars, last_speaker = initiate_swarm_chat(
+        initial_agent=agent1, messages=multiple_messages, agents=[agent1, agent2]
+    )
+
+    assert isinstance(last_speaker, SwarmAgent)
```
Comment (yiranwu0) [inline test/agentchat/contrib/test_swarm.py:162]:
```suggestion
```

File: test/agentchat/contrib/test_swarm.py [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+        }
+
+    # Mock an LLM response by overriding the generate_oai_reply function
+    for agent in [agent1, agent2]:
+        for reply_func_tuple in agent._reply_func_list:
+            if reply_func_tuple["reply_func"].__name__ == "generate_oai_reply":
+                if agent == agent1:
+                    reply_func_tuple["reply_func"] = mock_generate_oai_reply
+                elif agent == agent2:
+                    reply_func_tuple["reply_func"] = mock_generate_oai_reply_tool
```
Comment (yiranwu0) [inline test/agentchat/contrib/test_swarm.py:356]: can we do 

```
agent1._reply_func_list.append(...)
agent2._reply_func_list.append(...)
```
Not sure why it needs a for loop

File: test/agentchat/contrib/test_swarm.py [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+        initial_agent=agent1, messages=TEST_MESSAGES, agents=[agent1, agent2]
+    )
+
+    # Verify no message has name "_User"
+    for message in chat_result.chat_history:
+        assert message.get("name") != "_User"
+
+
+def test_context_variables_updating():
+    """Test context variables handling in tool calls"""
```
Comment (yiranwu0) [inline test/agentchat/contrib/test_swarm.py:246]: I think we can remove this. Testing multiple tool call covers this test.

File: test/agentchat/contrib/test_swarm.py (hunk 6 above)
Comment (yiranwu0) [inline test/agentchat/contrib/test_swarm.py:356]: same as above

File: test/agentchat/contrib/test_swarm.py [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+    bad_agent = ConversableAgent("bad_agent")
+
+    with pytest.raises(AssertionError, match="Invalid After Work value"):
+        agent1.register_hand_off(hand_to=AFTER_WORK(bad_agent))
+
+    with pytest.raises(AssertionError, match="Invalid After Work value"):
+        agent1.register_hand_off(hand_to=AFTER_WORK(0))
+
+    with pytest.raises(AssertionError, match="Agent must be a SwarmAgent"):
+        agent1.register_hand_off(hand_to=ON_CONDITION(0, "Testing"))
```
Comment (yiranwu0) [inline test/agentchat/contrib/test_swarm.py:408]: Consider remove. The initialization logic is very simple so it is not worth testing another type to make sure it returns invalid

File: autogen/agentchat/contrib/swarm_agent.py (hunk 1 above)
Comment (marklysze) [reply autogen/agentchat/contrib/swarm_agent.py:377]: I tested this, unfortunately with multiple tool calls, it injects the values of the context variables before it calls them. Therefore, if the first tool call updates the context variables the second tool call still has the old value. So we need to update the tool call parameters after each iteration.

File: test/agentchat/contrib/test_swarm.py (hunk 2 above)
Comment (marklysze) [reply test/agentchat/contrib/test_swarm.py:85]: No problem, removed.

File: test/agentchat/contrib/test_swarm.py (hunk 3 above)
Comment (marklysze) [reply test/agentchat/contrib/test_swarm.py:32]: No problem, removed the first part

File: test/agentchat/contrib/test_swarm.py (hunk 4 above)
Comment (marklysze) [reply test/agentchat/contrib/test_swarm.py:144]: No, it's not used

File: test/agentchat/contrib/test_swarm.py (hunk 5 above)
Comment (marklysze) [reply test/agentchat/contrib/test_swarm.py:162]: Okay, updated

File: test/agentchat/contrib/test_swarm.py (hunk 6 above)
Comment (marklysze) [reply test/agentchat/contrib/test_swarm.py:356]: I've updated with registering the reply functions instead.

Comment (marklysze) [reply test/agentchat/contrib/test_swarm.py:356]: Yep, I've updated them all :)

**Comment by marklysze** [comment]
@yiranwu0, thanks so much for testing. Can you have another look over please.

File: test/agentchat/contrib/test_swarm.py (hunk 4 above)
Comment (yiranwu0) [reply test/agentchat/contrib/test_swarm.py:137]: I am not sure if this fits the `test_swarm_transitions` name, for me it basically tests that the first agent is always the initial agent, not testing the transitions. Maybe remove this, or change this to one that actually does transitions? (Using the mock transfer tool call)

File: test/agentchat/contrib/test_swarm.py [hunk 9]
```diff
@@ -0,0 +1,453 @@
...
+    # Test the chat
+    chat_result, context_vars, last_speaker = initiate_swarm_chat(
+        initial_agent=test_second_agent,
+        user_agent=test_user,
+        messages=messages_one_w_name,
+        agents=[test_initial_agent, test_second_agent],
+    )
+
+    assert chat_result.chat_history[0].get("name") == "MyUser"  # Should persist
+    assert chat_result.chat_history[1].get("name") == "SecondAgent"
```
Comment (yiranwu0) [inline test/agentchat/contrib/test_swarm.py:131]: How about we pass in a list of messages that has names from the agent list? We should add another to test the resume chat feature to see whether the first agent to initiate the chat is correct.

Comment (marklysze) [reply test/agentchat/contrib/test_swarm.py:131]: Thank @yiranwu0!

I'll change the test_swarm_transitions function name to `test_initial_agent`.

For the resume you noted, good idea... I'll do that in this test_initial_agent function as it relates to the initial agent.

File: test/agentchat/contrib/test_swarm.py (hunk 4 above)
Comment (marklysze) [reply test/agentchat/contrib/test_swarm.py:137]: Removed :)

File: test/agentchat/contrib/test_swarm.py (hunk 9 above)
Comment (marklysze) [reply test/agentchat/contrib/test_swarm.py:131]: I finally decided to do the resume test in a new test, `test_resume_speaker()`.

**Review by yiranwu0 — APPROVED** [review]

**Comment by marklysze** [comment]
Note: OpenAI tests passed, except for one that retrieves the latest news (which contained some restricted words for Azure and its content filter was triggered). So will merge on that basis.

**Pull request merged**